src/java.base/share/native/libjava/RCULock.c

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (c) 2023, Azul Systems, Inc. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/** \file */
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "jni.h"
+#include "jvm.h"
+#include "jni_util.h"
+#include "jdk_crac_RCULock.h"
+
+static jfieldID readerThreadsListField = NULL;
+static jfieldID readCriticalMethodsField = NULL;
+
+JNIEXPORT void JNICALL
+Java_jdk_crac_RCULock_initFieldOffsets(JNIEnv *env, jclass cls)
+{
+    readerThreadsListField = (*env)->GetFieldID(env, cls, "readerThreadsList", "J");
+    readCriticalMethodsField = (*env)->GetFieldID(env, cls, "readCriticalMethods", "J");
```

Comment:
Looking up fields correctly in JNI is a pain to get the error handling correct.  I prefer to do the lookup reflectively and then pass the j.l.reflect.Field objects into the JNI method and convert them using `ToReflectField` [0]

It moves all the tricky error handling into Java and makes for obviously correct code.

If we keep the existing code, we need to null check the `readerThreadsListField` value before attempting a second JNI call as we can't call in with an exception pending.

[0] https://docs.oracle.com/javase/7/docs/technotes/guides/jni/spec/functions.html#from_reflected_field